Scripts/Services/Pathing/FastAStarAlgorithm.cs

```diff
@@ -91,9 +91,6 @@ public override Direction[] Find(IPoint3D p, Map map, Point3D start, Point3D goa
 
                 MoveImpl.Goal = Point3D.Zero;
 
-                if (count == 0)
```

Comment:
This is shortcutting the if statement below it. So deleting it is a functional no-op.

Reply:
That would be the case, if it were "continue" instead of "break". With "break" we are exiting the while loop. This is equal to claiming that the open list ist empty (the condition fo the while loop), which isn't necessarily the case at this point.

Reply:
So it should be `continue`?

Reply:
I think it should be removed entirely. 

The point is the following: The AStar algorithm keeps a list of grid points ("nodes") called "open list", which it considers promising steps towards the goal. From these it repeatedly choses the most promising one to "expand", i.e. to remove from the list and then to check for the possibility of adding its neighbors to the open list. There should, in principle, be three ways the algorithm can terminate:

1) The expansion of the node encounters the goal, in which case the algorithm has succeeded.
2) The number of expanded nodes reaches some maximum value which was set as an upper limit, in which case the algorithm has failed.
3) There are no more nodes left on the open list, i.e. all expansions have either run into barriers or the edge of the grid, in case of which the algorithm has failed. 

The outermost while-loop here `while (m_OpenList != -1){ ... }` tests condition number 3). 

What the if-statement does, is terminate the algorithm with failure if the expansion of *any* node fails, i.e. is blocked from all directions, even if there are other nodes left on the open list. This is not what should happen. The correct thing to do is to move on to the next best node on the open list and expand that one. 

So yes, if you will, replacing "break" by "continue" would be correct. But then the statement would be redundant and could be dropped. 

The problem with having the break statement in is that, if some node expands into a "pocket" surrounded by barriers, and the node inside the pocket gets chosen next for expansion, then the algorithm will terminate, even if there is a simple way to go around the obstacle. 

In some tests that I ran, this happened for instance when trying to exit a cluttered building through a narrow door. Deleting the if-statement fixed the problem and allowed the algorithm to succeed even in such cases.